### PR TITLE
[aot_autograd] Reduce runtime_wrapper Python overhead: 4 micro-optimizations

### LINKED ISSUE #161783: torch.compile has runtime overhead on small graphs
Reported by @BoyuanFeng.

One of the use cases for torch.compile is as a "kernel generator".  When we try to use torch.compile to generate a kernel for rope, we see significant CPU overhead. ([run via tritonbench](https://github.com/meta-pytorch/tritonbench/blob/main/run.py)).

We should investigate where this runtime overhead is coming from and squash it.

Image

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @amjames @Lucaskabela @jataylo @bdhirsh @bobrenjc93 @aorenste @chenyang78

### PR DESCRIPTION
**Pull request by raviteja-ganta**
## Summary

\`torch.compile\`'d functions pay a Python-layer overhead on every call through
\`runtime_wrapper\` / \`_codegen_runtime_wrapper\`, even when the compiled kernel
itself is fast. This PR applies four targeted micro-optimizations that cut
those allocations.

**Fix A — Skip profiler closure when profiling is off**
\`runtime_wrapper\` previously created an \`exit_prologue\` closure + two
\`nonlocal\` cells on every single call, even when the profiler was completely
disabled. The closure is now only created when
\`torch.autograd.profiler._is_profiler_enabled\` is \`True\`. The common
(non-profiling) path takes the fast \`else\` branch directly.

**Fix B — Module-level \`nullcontext()\` singleton**
\`_FirstInvocationContext.__call__\` returned a freshly allocated
\`nullcontext()\` on every call after the first invocation. We now return a
module-level \`_NOOP_CTX = nullcontext()\` singleton, eliminating one object
allocation per call.

**Fix C — Skip empty \`orig_inputs\` dict allocation**
\`_codegen_capture_orig_inputs\` emitted \`orig_inputs = {}\` into the generated
wrapper even when \`epilogue_args_idx\` is empty (no mutated inputs). That dead
dict was allocated and immediately discarded on every inference call. We now
only emit \`orig_inputs = {...}\` when there are actually indices to capture.

**Fix D — Direct-return fast path for trivial wrappers**
For inference functions with no mutations, no aliased outputs, no dynamic
outputs, and no grad-enabled mutation (\`is_trivial=True\`), the generated
\`_runtime_wrapper\` collapses to:

```python
return _normalize_as_list_(_compiled_fn_(args))
```

eliminating the output-arity assertion and two dead variable assignments
(\`fw_outs\`, \`ret_outs\`). \`normalize_as_list\` is retained to preserve the
list-return contract that the downstream dispatcher depends on.

## Benchmarks

Measured using \`benchmarks/dynamo/microbenchmarks/overheads.py\` (the same
benchmark used in #163747), built from source on Apple Silicon (CPU-only):

| Scenario | Before | After | Saved |
|---|---|---|---|
| \`requires_grad=False\` | 9.8 µs | 9.6 µs | 0.2 µs (2%) |
| \`inference_mode()\` | 9.5 µs | 8.8 µs | **0.7 µs (7%)** |
| \`requires_grad=True\` | 18.8 µs | 18.5 µs | within noise |

Fixes A+B+C apply to all scenarios. Fix D applies to inference paths
(no grad, no mutations, no aliasing). The \`requires_grad=True\` case benefits
less because Fix D does not apply when \`trace_joint=True\`.

This is the same class of fix as #163747 (which applied a similar profiler
fast-path to \`output_code.py\`).

## Testing

- Smoke tests covering: trivial inference (single + multi-output), in-place
  mutation (epilogue fires correctly), output aliasing, profiler slow path
  (RecordFunction fires when profiling on), and training/backward pass.
- \`is_trivial=False\` for any non-trivial case ensures the full epilogue path
  is unchanged.
- \`test_open_device_registration\` (extension backend / privateuseone device)
  verified passing — this test exercises the exact trivial-inference path that
  Fix D rewrites.

Fixes #161783

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review these changes

**Comment by zou3519** [comment]
deferring to @aorenste

File: test/functorch/test_codegen_runtime_wrapper.py [hunk 1]
```diff
@@ -219,18 +220,24 @@ def test_output_arity_validation_baked(self):
         """
         The expected output arity should be baked into the generated code
         as a constant, not computed at runtime.
+
+        The function returns an aliased output (x.view(-1)) to ensure the
+        non-trivial epilogue path is taken (is_trivial=False). The compiled
+        function returns 3 non-aliased outputs; the aliased view is
+        reconstructed by _replay_aliases_ in the epilogue. The arity check
+        `if len(all_outs) != 3:` verifies that 3 is baked as a constant.
```
Comment (aorenste) [inline test/functorch/test_codegen_runtime_wrapper.py]: [cleanup] This docstring contradicts the assertion below it. It says the compiled function "returns 3 non-aliased outputs" and that the arity check is `if len(all_outs) != 3:`, but the function returns 4 outputs (`x+1, x*2, x-1, x.view(-1)`) and line 240 asserts `!= 4`. The baked arity counts all user outputs (the aliased view still occupies a slot in `all_outs` before `_replay_aliases_` reconstructs it), so 4 is correct and the docstring is stale -- a leftover from the pre-`x.view(-1)` 3-output version. Update the docstring to 4 (and reword "returns 3 non-aliased outputs" so it does not read as the total arity).

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 2]
```diff
@@ -831,7 +853,8 @@ def _codegen_compiled_fn_invocation(
             )
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: This can be a lot more shared... something like:
```suggestion
        output = "return" if is_trivial else "all_outs ="
        if disable_amp:
            rw_globals["_DisableAutocast_"] = torch._C._DisableAutocast
            rw_lines.append("            with _DisableAutocast_():")
            rw_lines.append(
                f"                {output} _normalize_as_list_(_compiled_fn_(args))"
            )
        else:
            rw_lines.append(
                f"            {output} _normalize_as_list_(_compiled_fn_(args))"
            )
```

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 3]
```diff
@@ -786,6 +796,7 @@ def _codegen_compiled_fn_invocation(
     trace_joint: bool,
     indices_of_inps_to_detach: list[int],
     disable_amp: bool,
+    is_trivial: bool = False,
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: [cleanup] `is_trivial` is only ever called from the single site in `_create_runtime_wrapper`, which always passes it explicitly, so this `= False` default is never exercised (the identical default on `_codegen_epilogue` at line 867 has the same single caller). Make `is_trivial` a required parameter on both -- a future caller that forgets it then fails loudly with a TypeError instead of silently falling back to the slower non-trivial path.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 4]
```diff
@@ -1113,24 +1136,38 @@ def record_runtime_wrapper_prologue_exit(
 
     @simple_wraps(_inner_compiled_fn)
     def runtime_wrapper(args: list[Any]) -> Any:
-        cm = record_runtime_wrapper_prologue_enter()
-        prologue_exited = False
+        if (
+            torch.autograd.profiler._is_profiler_enabled
+            and dynamo_config.record_runtime_overhead
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:1190]: I'm tempted to say we should specialize on `record_runtime_overhead` (return a different version of runtime_wrapper based on it) but it defaults to True so that wouldn't save basically anything for the default case.

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

Close to landing. Two things to address before merge, plus minor nits.

[blocker] Add a regression test for the trivial fast-path with multiple outputs: compile a multi-output inference function with no aliases or mutations and assert the generated source uses `return _normalize_as_list_(_compiled_fn_(args))` (and contains no `all_outs`), and that the returned value preserves the list contract for 2+ outputs. This guards the list-vs-tuple return contract the direct-return path had to restore.

[cleanup] Collapse the duplicated trivial / disable_amp / plain emit branches in `_codegen_compiled_fn_invocation` by parameterizing on the assignment target -- see the inline suggestion on the `is_trivial` block.

Remaining inline nits: the stale docstring (`!= 3` vs `!= 4`) and the `is_trivial=False` defaults on the codegen helpers. The PR description also lists a profiler-on smoke test that is not in the diff.

**Comment by raviteja-ganta** [comment]
Thanks @aorenste for the thorough review. Addressed all points, and also rebased onto latest main to clear a merge conflict that arose from the AOT output-normalization / view-replay inlining PRs ([#184384](https://github.com/pytorch/pytorch/pull/184384), [#184390](https://github.com/pytorch/pytorch/pull/184390)) landing in parallel.

[blocker] Multi-output trivial-path regression test - added test_inference_trivial_multi_output, which compiles a 3-output inference fn and asserts the generated source returns the normalized list directly (no arity check / epilogue) while preserving the multi-output contract.

[cleanup] Dedup the trivial/disable_amp branches - done. Note the rebase onto the new inlined-normalization structure reshaped this region: the trivial path is now all_outs = _compiled_fn_(args) -> _codegen_normalize_as_list(...) -> return all_outs, so the duplication is naturally collapsed and the is_trivial branch is just a one-line early return.

[nit] Stale != 3 docstring - fixed to 4, reworded so the non-aliased count isn't read as total arity.

[nit] is_trivial=False defaults - removed; is_trivial is now a required parameter on both _codegen_compiled_fn_invocation and _codegen_epilogue.

Profiler-on smoke test - added test_profiler_prologue_fires_when_profiling, which enables the profiler and asserts the AOTDispatcher Runtime Wrapper Prologue RecordFunction fires (the Fix A slow path), so the PR description claim now matches the diff.

File: test/functorch/test_codegen_runtime_wrapper.py (hunk 1 above)
Comment (raviteja-ganta) [reply test/functorch/test_codegen_runtime_wrapper.py]: Done - updated the docstring to 4 and reworded so the "3 plain outputs" no longer reads as the total arity.

File: test/functorch/test_codegen_runtime_wrapper.py [hunk 5]
```diff
@@ -45,12 +46,60 @@ def f(x):
...
+            x = torch.randn(4)
+            out = f(x)
+
+        self.assertEqual(out, (x + 1, x * 2, x - 3))
+        self.assertEqual(len(out), 3)
+        self.assertEqual(len(captured), 1)
+        source = captured[0]
+        self.assertIn("all_outs = _compiled_fn_(args)", source)
+        self.assertIn("return all_outs", source)
+        self.assertNotIn("if len(all_outs) !=", source)
```
Comment (aorenste) [inline test/functorch/test_codegen_runtime_wrapper.py:78]: [cleanup] This test was added to guard the list-vs-tuple return contract the trivial path had to restore -- the `_codegen_normalize_as_list` step that fixed `test_open_device_registration`'s "NoneType not subscriptable". But the assertions here don't pin that contract: `all_outs = _compiled_fn_(args)`, `return all_outs`, and the absent arity check all stay true even if normalization is dropped from the trivial path. The behavioral `assertEqual(out, (...))` can't catch it either -- Dynamo reconstructs the tuple at the user level regardless of whether `all_outs` is a list or tuple internally. To actually guard it, assert the normalization is emitted, e.g. `self.assertIn("all_outs = list(all_outs)", source)`, or add a single-output / None-returning case that would fail without normalization.

File: test/functorch/test_codegen_runtime_wrapper.py [hunk 6]
```diff
@@ -45,12 +46,60 @@ def f(x):
...
+
+    def test_profiler_prologue_fires_when_profiling(self):
```
Comment (aorenste) [inline test/functorch/test_codegen_runtime_wrapper.py:85]: [cleanup] This fires the prologue RecordFunction only because `dynamo_config.record_runtime_overhead` happens to default to True (torch/_dynamo/config.py:890) -- the Fix A guard is `_is_profiler_enabled and record_runtime_overhead`. The test doesn't pin it, so if that default ever flips the test fails for an unrelated reason. Make the intent explicit and robust with `@torch._dynamo.config.patch(record_runtime_overhead=True)` on the method.

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

File: test/functorch/test_codegen_runtime_wrapper.py (hunk 6 above)
Comment (raviteja-ganta) [reply test/functorch/test_codegen_runtime_wrapper.py:85]: Done - added @torch._dynamo.config.patch(record_runtime_overhead=True) to the method so it no longer relies on the default. The test now pins both halves of the Fix A guard explicitly.

File: test/functorch/test_codegen_runtime_wrapper.py (hunk 5 above)
Comment (raviteja-ganta) [reply test/functorch/test_codegen_runtime_wrapper.py:78]: Good catch - you're right the old assertions stayed true even without normalization. Added assertions on the normalization the trivial path actually emits:

self.assertIn("if isinstance(all_outs, tuple):", source)
self.assertIn("elif not isinstance(all_outs, list):", source)

These come only from _codegen_normalize_as_list, so the test now fails if normalization is dropped from the trivial path. (FWIW the literal all_outs = list(all_outs) is emitted as the indented body of that isinstance block rather than a standalone line, so I asserted on the discriminating if/elif lines instead.) I skipped the None-returning case since a None return through @torch.compile produces zero graph outputs and wouldn't hit the multi-output trivial path.
